README.md
ldmicro-jp
==========

Ladder Logic for PIC and AVR ( http://cq.cx/ladder.pl ).
This fork from Release 2.2 is for Japanese translation.

Credits
=======

LDmicro was created by Jonathan Westhues.

_August 2010, New York City; thanks to G. Mariani and A. Pedrazzi for Italian translations; H. U. Noell for German; D. Corteletti and O. da Rocha for Portuguese; M. Vaufleury for French; J. Pascual for Spanish; A. Akici for Turkish._



ldmicro/CHANGES.txt

== Release 2.3

    * Fix buffer overrun or write to null pointer if Windows sends an
      LVN_GETDISPINFO without a valid item.pszText, which happens now
      under Win10.

== Release 2.2

    * Fix a problem with the ANSI C target when the program had bit and
      integer variables with the same name. Note that this changes the
      names of the symbols in the generated C program; so a system that
      uses "magic variables" with this target for I/O must be updated
      to use the new names.

    * Fix a subtle bug in the PIC16 add and subtract routines, where 
      operations of the form B = A - B could fail.

    * The piecewise linear tables were broken for the AVRs; fix that.

== Release 2.1

    * For the AVR UARTs, poll UDRE instead of TXC. The formatted string op
      was broken on some targets, should now be fixed.

    * Don't draw selected op in bold font; that looks ugly under Vista.

== Release 2.0
    
    * Add PIC16F886 and PIC16F887 targets.

    * Fix display bug in the list to select an I/O pin.

    * Fix bug where PIC16 UART locks up forever after a framing error when
      the cycle time is faster than one byte time.

    * Fix bug where PIC16 outputs could briefly glitch high at startup.

    * Clear PCLATH in PIC16 boot vector, since some bootloaders expect that.

== Release 1.9

    * Modify PIC16 boot vectors to work with many bootloaders.

== Release 1.8

    * Fix modification of a constant string that blew up in new MSVC++
      compiler.

    * Add Italian, Turkish, Portuguese.

== Release 1.7

    * Make the source compile with latest version of MSVC++; overloaded
      functions behave a bit differently.

    * Recover from (and ignore) UART errors on the PIC16 target, instead
      of getting stuck forever.

    * Whenever contacts bound to an output pin (Yfoo) were edited, they
      reverted to an input pin (Xfoo); now fixed.

    * Don't abort on too-wide program; instead display nice message.

    * It was possible (by adding and deleting contacts/coils with the
      same name) to end up with two bit variables bound to the same
      physical I/O pin; now fixed.

    * File -> Open was correct, but Ctrl+O failed to ask about unsaved
      changes before opening requested file; now both are correct.

    * Add Spanish user interface strings.

== Release 1.6

    * Internationalize the user interface strings; we now have versions
      in English, French, and German.

    * First source release, under the GPLv3.

== Release 1.5

    * Add untested support for ATmega32.

    * Remove annoying lag in user interface when editing large (hundreds
      of ops) programs

== Release 1.4

    * Fix a terrible bug in the target for the ATmega8; because there is
      no PORTA/DDRA/PINA, I broke an assumption in my code and failed
      to set up the port directions.

== Release 1.3

    * Timer delays are represented as a signed 32-bit integer count
      of microseconds. If the user provides a delay >= 2**31 us, then
      show an error instead of just letting things wrap.

    * Change the start-up behaviour of TOF timers. Previously they would
      start from a count of zero, so they would be on (independent of
      rung-in) until they counted themselves off. Now they start out
      at full count (as if rung-in has been low for a very long time),
      so rung-out is low until rung-in goes high.

== Release 1.2

    * Add an untested target for the ATmega8

    * Add a special instruction to simplify piecewise linear tables

    * Fix some user interface bugs: it was possible to drag the top of the
      I/O list so high that you couldn't grab it again, and there were
      some cases in which the pin number associated with UART and PWM
      variables was not displayed

== Release 1.1

    * Fix persistent variables, which were broken for the PIC16F628

== Release 1.0

    * Fix bug in which the filename that appears in the title bar of the
      main window failed to get updated when opening/saving a file using
      the keyboard shortcuts (Ctrl+O/+S)

    * Fix simulation crash when the ladder logic program divides by zero

    * Fix jumpy scrolling on programs with many rungs of logic when the
      cursor is off-screen

== Release 0.9

    * Fix bug with formatted string op on the AVR
    * Fix previously-untested ATmega16 and ATmega162 targets, which were
      completely broken

=== Release 0.8

    * Fix PORTA on the PIC16F819 (came up assigned to ADCs, of course)

=== Release 0.7

    * Support arbitrary character (\xAB) escapes in formatted string op
    * Fix a bug in which the title bar of the main window was not updated

=== Release 0.6

    * Add formatted text output over serial (e.g. to an LCD or a PC)
    * Add ability to make variables persistent (i.e. auto-saved in EEPROM)
    * Add look-up table instructions
    * Fix a bug with the PORTE pins on some AVRs
    * Fix miscellaneous user interface bugs

=== Release 0.5

    * Interpretable byte code target
    * Shift register and master control relay instructions

=== Release 0.4

    * Make ADCs work on the AVRs

=== Release 0.3

    * Support serial for AVR
    * Support PWM for PIC16 and AVR
    * Show program filename in title bar of main window
    * Untested support for PIC16F88, F819, F876
    * Generate ANSI C code from ladder diagram

=== Release 0.2

    * Support serial communications (using UART), PIC16 only
    * Support ADC reads, PIC16 only
    * Simulation environment for ADC and serial
    * Support ASCII character constant ('a') literals
    * Fix PORTA pins in PIC16F628 (should assign as GPIO, not to comparator)
    * Make file open/save dialogs work under Win98
    * Fix PORTA/PORTE pins in PIC16F877 (should assign as GPIO, not to ADC)
    * Add ability to comment your program
    * Fix bug when a relative filename is given on the command line and
      the `Compile As' dialog is later used to specify a destination in
      a different directory


ldmicro/lang.cpp
//-----------------------------------------------------------------------------
// Copyright 2007 Jonathan Westhues
//
// This file is part of LDmicro.
// 
// LDmicro is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
// 
// LDmicro is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
// 
// You should have received a copy of the GNU General Public License
// along with LDmicro.  If not, see <http://www.gnu.org/licenses/>.
//------
//
// Multiple language support. For every non-English language, we have a 
// table that maps the English strings to the translated strings. An
// internationalized version of the program will attempt to translate any
// string that is passed, using the appropriate (selected by a #define)
// table. If it fails then it just returns the English.
// Jonathan Westhues, Apr 2007
//-----------------------------------------------------------------------------
#include <windows.h>
#include <stdio.h>
#include <stdlib.h>

#include "ldmicro.h"

typedef struct LangTableTag {
    char *from;
    char *to;
} LangTable;

typedef struct LangTag {
    LangTable   *tab;
    int          n;
} Lang;

// These are the actual translation tables, so should be included in just
// one place.
#include "obj/lang-tables.h"

char *_(char *in)
{
    Lang *l;

#if defined(LDLANG_EN)
    return in;
#elif defined(LDLANG_DE)
    l = &LangDe;
#elif defined(LDLANG_FR)
    l = &LangFr;
#elif defined(LDLANG_ES)
    l = &LangEs;
#elif defined(LDLANG_IT)
    l = &LangIt;
#elif defined(LDLANG_TR)
    l = &LangTr;
#elif defined(LDLANG_PT)
    l = &LangPt;
#elif defined(LDLANG_JA)
    l = &LangJa;
#else
#   error "Unrecognized language!"
#endif

    int i;

    for(i = 0; i < l->n; i++) {
        if(strcmp(in, l->tab[i].from)==0) {
            return l->tab[i].to;
        }
    }

    return in;
}


ldmicro/txt2c.pl
#!/usr/bin/perl

use strict;

print <<EOT;
// generated by txt2c.pl from $ARGV[0]
#include <stdlib.h>
EOT

my %manuals = map { /manual-?([^.]*)/ or die; (lc($1) || 'en') => $_ } glob('manual*txt');
my @languages_without_manual = grep { !exists $manuals{$_} } map { /-([^.]*)/ or die; lc $1 } glob('lang-*.txt');

for my $lang (sort keys %manuals) {
    my ($name, $ifdef);
    my $manual = $manuals{$lang};

    if ($lang eq 'en') {
        # Some languages don't have translated manuals yet, so use English
        my @langs = sort('en', @languages_without_manual);
        my $cond = join ' || ', map { "defined(LDLANG_$_)" } map { uc } @langs;
        
        $name = "HelpText";
        $ifdef = "#if $cond";
    } else {
        $name = "HelpText" . ucfirst $lang;
        $ifdef = "#ifdef LDLANG_" . uc $lang;
    }

    print <<EOT;
$ifdef
char *$name\[] = {
EOT

    open(IN, $manual) or die;
    while(<IN>) {
        chomp;
        # avoid problem with Japanese encoding (cp932) which has 0x5c in a multibyte character
        s/\\/\\\\/g if $lang ne 'ja';
        s/"/\\"/g;

        print qq{    "$_",\n};
    }
    close IN;

    print <<EOT;
    NULL
};
#endif

EOT

}


ldmicro/lang-tables.pl
#!/usr/bin/perl

use strict;

my $t;

for my $file (sort <lang-*.txt>) {

    my $name = $file;
    $name =~ s#lang-##;
    $name =~ s#(.)#uc($1)#e;
    $name =~ s#\.txt##;
    $nameUc = uc($name);
    
    print "#ifdef LDLANG_$nameUc\n";
    print "static LangTable Lang${name}Table[] = {\n";

    my $engl = 1;
    local $. = 0;
    open(my $in, '<', $file) or die "couldn't open $file";
    while(<$in>) {
        chomp;

        if (/^\s*$/) {
            if ($engl) {
                next;
            }
            die "blank line mid-translation at $file, $.\n";
        }

        if ($engl) {
            $toTranslate = $_;
            $engl = 0;
        } else {
            $translated = $_;

            print "    { $toTranslate, $translated },\n";
            $engl = 1;
        }
    }

    print "};\n";

    print <<EOT;
static Lang Lang$name = {
    Lang${name}Table, sizeof(Lang${name}Table)/sizeof(Lang${name}Table[0])
};
#endif
EOT

    close FILE;
}



ldmicro/helpdialog.cpp
//-----------------------------------------------------------------------------
// Copyright 2007 Jonathan Westhues
//
// This file is part of LDmicro.
// 
// LDmicro is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
// 
// LDmicro is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
// 
// You should have received a copy of the GNU General Public License
// along with LDmicro.  If not, see <http://www.gnu.org/licenses/>.
//------
//
// Window to show our internal help. Roll my own using a rich edit control so
// that I don't have to distribute a separate Windows Help or HTML file; the
// manual is compiled in to the exe. Do pretty syntax highlighting style
// colours.
// Jonathan Westhues, Dec 2004
//-----------------------------------------------------------------------------
#include <windows.h>
#include <stdio.h>
#include <stdlib.h>
#include <commctrl.h>
#include <richedit.h>

#include "ldmicro.h"

extern char *HelpText[];
extern char *HelpTextDe[];
extern char *HelpTextFr[];
extern char *HelpTextTr[];
extern char *HelpTextJa[];

static char *AboutText[] = {
"",
"ABOUT LDMICRO",
"=============",
"",
"LDmicro is a ladder logic editor, simulator and compiler for 8-bit",
"microcontrollers. It can generate native code for Atmel AVR and Microchip",
"PIC16 CPUs from a ladder diagram.",
"",
"This program is free software: you can redistribute it and/or modify it",
"under the terms of the GNU General Public License as published by the",
"Free Software Foundation, either version 3 of the License, or (at your",
"option) any later version.",
"",
"This program is distributed in the hope that it will be useful, but",
"WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY",
"or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License",
"for more details.",
"",
"You should have received a copy of the GNU General Public License along",
"with this program. If not, see <http://www.gnu.org/licenses/>.",
"",
"The source code for LDmicro is available at",
"",
"    http://cq.cx/ladder.pl",
"",
"Copyright 2005-2016 Jonathan Westhues",
"Release 2.3, built " __TIME__ " " __DATE__ ".",
"",
"Email: user jwesthues, at host cq.cx",
"",
NULL
};

static char **Text[] = {
#if defined(LDLANG_EN) || \
    defined(LDLANG_ES) || \
    defined(LDLANG_IT) || \
    defined(LDLANG_PT)
    HelpText,
#elif defined(LDLANG_DE)
    HelpTextDe,
#elif defined(LDLANG_FR)
    HelpTextFr,
#elif defined(LDLANG_TR)
    HelpTextTr,
#elif defined(LDLANG_JA)
    HelpTextJa,
#else
#   error "Bad language"
#endif

    // Let's always keep the about text in English.
    AboutText
};

static HWND HelpDialog[2];
static HWND RichEdit[2];

static BOOL HelpWindowOpen[2];

static int TitleHeight;

#define RICH_EDIT_HEIGHT(h)  \
    ((((h) - 3 + (FONT_HEIGHT/2)) / FONT_HEIGHT) * FONT_HEIGHT)

static void SizeRichEdit(int a)
{
    RECT r;
    GetClientRect(HelpDialog[a], &r);

    SetWindowPos(RichEdit[a], HWND_TOP, 6, 3, r.right - 6, 
        RICH_EDIT_HEIGHT(r.bottom), 0);
}

static BOOL Resizing(RECT *r, int wParam)
{
    BOOL touched = FALSE;
    if(r->right - r->left < 650) {
        int diff = 650 - (r->right - r->left);
        if(wParam == WMSZ_RIGHT || wParam == WMSZ_TOPRIGHT ||
            wParam == WMSZ_BOTTOMRIGHT)
        {
            r->right += diff;
        } else {
            r->left -= diff;
        }
        touched = TRUE;
    }

    if(!(wParam == WMSZ_LEFT || wParam == WMSZ_RIGHT)) {
        int h = r->bottom - r->top - TitleHeight - 5;
        if(RICH_EDIT_HEIGHT(h) != h) {
            int diff = h - RICH_EDIT_HEIGHT(h);
            if(wParam == WMSZ_TOP || wParam == WMSZ_TOPRIGHT ||
                wParam == WMSZ_TOPLEFT)
            {
                r->top += diff;
            } else {
                r->bottom -= diff;
            }
            touched = TRUE;
        }
    }

    return !touched;
}

static void MakeControls(int a)
{
    HMODULE re = LoadLibrary("RichEd20.dll");
    if(!re) oops();

    RichEdit[a] = CreateWindowEx(0, RICHEDIT_CLASS,
        "", WS_CHILD | WS_CLIPSIBLINGS | WS_VISIBLE | ES_READONLY |
        ES_MULTILINE | WS_VSCROLL,
        0, 0, 100, 100, HelpDialog[a], NULL, Instance, NULL);

    SendMessage(RichEdit[a], WM_SETFONT, (WPARAM)FixedWidthFont, TRUE);
    SendMessage(RichEdit[a], EM_SETBKGNDCOLOR, (WPARAM)0, RGB(0, 0, 0));

    SizeRichEdit(a);

    int i;
    BOOL nextSubHead = FALSE;
    for(i = 0; Text[a][i]; i++) {
        char *s = Text[a][i];

        CHARFORMAT cf;
        cf.cbSize = sizeof(cf);
        cf.dwMask = CFM_BOLD | CFM_COLOR;
        cf.dwEffects = 0;
        if((s[0] == '=') ||
           (Text[a][i+1] && Text[a][i+1][0] == '='))
        {
            cf.crTextColor = RGB(255, 255, 110);
        } else if(s[3] == '|' && s[4] == '|') {
            cf.crTextColor = RGB(255, 110, 255);
        } else if(s[0] == '>' || nextSubHead) {
            // Need to make a copy because the strings we are passed aren't
            // mutable.
            char copy[1024];
            if(strlen(s) >= sizeof(copy)) oops();
            strcpy(copy, s);

            int j;
            for(j = 1; copy[j]; j++) {
                if(copy[j] == ' ' && copy[j-1] == ' ')
                    break;
            }
            BOOL justHeading = (copy[j] == '\0');
            copy[j] = '\0';
            cf.crTextColor = RGB(110, 255, 110);
            SendMessage(RichEdit[a], EM_SETCHARFORMAT, SCF_SELECTION,
                (LPARAM)&cf);
            SendMessage(RichEdit[a], EM_REPLACESEL, (WPARAM)FALSE,
                (LPARAM)copy);
            SendMessage(RichEdit[a], EM_SETSEL, (WPARAM)-1, (LPARAM)-1);
            
            // Special case if there's nothing except title on the line
            if(!justHeading) {
                copy[j] = ' ';
            }
            s += j;
            cf.crTextColor = RGB(255, 110, 255);
            nextSubHead = !nextSubHead;
        } else {
            cf.crTextColor = RGB(255, 255, 255);
        }

        SendMessage(RichEdit[a], EM_SETCHARFORMAT, SCF_SELECTION, (LPARAM)&cf);
        SendMessage(RichEdit[a], EM_REPLACESEL, (WPARAM)FALSE, (LPARAM)s);
        SendMessage(RichEdit[a], EM_SETSEL, (WPARAM)-1, (LPARAM)-1);

        if(Text[a][i+1]) {
            SendMessage(RichEdit[a], EM_REPLACESEL, FALSE, (LPARAM)"\r\n");
            SendMessage(RichEdit[a], EM_SETSEL, (WPARAM)-1, (LPARAM)-1);
        }
    }

    SendMessage(RichEdit[a], EM_SETSEL, (WPARAM)0, (LPARAM)0);
}

//-----------------------------------------------------------------------------
// Window proc for the help dialog.
//-----------------------------------------------------------------------------
static LRESULT CALLBACK HelpProc(HWND hwnd, UINT msg, WPARAM wParam,
    LPARAM lParam)
{
    int a = (hwnd == HelpDialog[0] ? 0 : 1);
    switch (msg) {
        case WM_SIZING: {
            RECT *r = (RECT *)lParam;
            return Resizing(r, wParam);
            break;
        }
        case WM_SIZE:
            SizeRichEdit(a);
            break;

        case WM_ACTIVATE:
        case WM_KEYDOWN:
            SetFocus(RichEdit[a]);
            break;
    
        case WM_DESTROY:
        case WM_CLOSE:
            HelpWindowOpen[a] = FALSE;
            // fall through
        default:
            return DefWindowProc(hwnd, msg, wParam, lParam);
    }

    return 1;
}

//-----------------------------------------------------------------------------
// Create the class for the help window.
//-----------------------------------------------------------------------------
static void MakeClass(void)
{
    WNDCLASSEX wc;
    memset(&wc, 0, sizeof(wc));
    wc.cbSize = sizeof(wc);

    wc.style            = CS_BYTEALIGNCLIENT | CS_BYTEALIGNWINDOW | CS_OWNDC |
                          CS_DBLCLKS;
    wc.lpfnWndProc      = (WNDPROC)HelpProc;
    wc.hInstance        = Instance;
    wc.hbrBackground    = (HBRUSH)GetStockObject(BLACK_BRUSH);
    wc.lpszClassName    = "LDmicroHelp";
    wc.lpszMenuName     = NULL;
    wc.hCursor          = LoadCursor(NULL, IDC_ARROW);
    wc.hIcon            = (HICON)LoadImage(Instance, MAKEINTRESOURCE(4000),
                            IMAGE_ICON, 32, 32, 0);
    wc.hIconSm          = (HICON)LoadImage(Instance, MAKEINTRESOURCE(4000),
                            IMAGE_ICON, 16, 16, 0);

    RegisterClassEx(&wc);
}

void ShowHelpDialog(BOOL about)
{
    int a = about ? 1 : 0;
    if(HelpWindowOpen[a]) {
        SetForegroundWindow(HelpDialog[a]);
        return;
    }

    MakeClass();

    char *s = about ? "About LDmicro" : "LDmicro Help";
    HelpDialog[a] = CreateWindowEx(0, "LDmicroHelp", s,
        WS_OVERLAPPED | WS_SYSMENU | WS_THICKFRAME | WS_MINIMIZEBOX |
        WS_SIZEBOX,
        100, 100, 650, 300+10*FONT_HEIGHT, NULL, NULL, Instance, NULL);
    MakeControls(a);
   
    ShowWindow(HelpDialog[a], TRUE);
    SetFocus(RichEdit[a]);

    HelpWindowOpen[a] = TRUE;

    RECT r;
    GetClientRect(HelpDialog[a], &r);
    TitleHeight = 300 - r.bottom;

    GetWindowRect(HelpDialog[a], &r);
    Resizing(&r, WMSZ_TOP);
    SetWindowPos(HelpDialog[a], HWND_TOP, r.left, r.top, r.right - r.left, 
        r.bottom - r.top, 0);
}
